bionic/deriver.py

```diff
@@ -369,16 +369,23 @@ class TaskKeyLogger:
     of a computation.)
     """
 
-    def __init__(self, level):
-        self._level = level
-        self._already_logged_task_keys = set()
+    def __init__(self, bootstrap):
+        self._level = logging.INFO if bootstrap is not None else logging.DEBUG
+
+        if bootstrap is not None and bootstrap.process_manager is not None:
+            already_logged_task_keys = bootstrap.process_manager.dict()
+        else:
+            already_logged_task_keys = {}
+        self._already_logged_task_keys = already_logged_task_keys
 
     def _log(self, template, task_key, is_resolved=True):
+        # TODO: We probably need to take a lock here for the case where multiple
```

Comment:
I agree that we should get rid of this eventually. But for now, I don't know how we can create a set-like class you described that is process safe. `Multiprocessing` doesn't offer any such class for us.

What I can do is make this a note so we know there is a potential for double logging (it's a very edge case scenario though). Wdyt?

Reply:
Couldn't we make our own custom class with a single `try_to_add` method and have it managed with a proxy?

I do think it's worth removing any race conditions like this because they could lead to a flaky test.

Reply:
Talked about this outside github, we do need a lock to implement `try_to_add` method.